### PR TITLE
[WIP/DMN] OpenACC forall and forallN

### PR DESCRIPTION
**Pull request by willkill07**
This is a work in progress.

1. PGI 17.4 breaks in the EDG frontend. I'm working on submitting a bug report to them
2. Due to the additional amount of directives offered, I wanted to offer an (arguably) cleaner way of specifying options. I opted for a `config<>` list that can be added to any `openacc` policy.
3. Due to PGI breaking and me still wanting to test this, I opted for using C++14 `[[deprecated("reason")]]` attributes on all SFINAE functions for various policy configurations. This works well and is by default disabled
5. `forallN` and `forall` machinery is all in place and works.
6. Users can specify `kernels` or `parallel` constructs

Snippet from `test/unit/acc/test.cpp` below:

```cpp
#define RAJA_ENABLE_VERBOSE
#define RAJA_ENABLE_OPENACC 1
#include "RAJA/policy/openacc.hpp"
#include "RAJA/pattern/forall.hpp"
#include "RAJA/pattern/forallN.hpp"


  const int N = 1024;
  using T = double;
  T *a, *b, *c;
  T* a = new T[N];
  T* b = new T[N];
  T* c = new T[N];

  namespace acc = RAJA::acc;

  // example 1 : forall

  RAJA::forall<
    RAJA::acc_kernels_loop_exec<
      acc::config<acc::NumGangs<32>, acc::NumVectors<32>>,
      acc::config<acc::Independent, acc::Gang, acc::Vector>>>
    (0, N, [=] (int i) {
    c[i] = a[i] + b[i];
  });

  // example 2 : forallN

  RAJA::forallN<
    RAJA::NestedPolicy<
      RAJA::ExecList<
        RAJA::acc_loop_exec<acc::config<acc::Independent>>,
        RAJA::acc_loop_exec<acc::config<acc::Independent>>>,
      RAJA::ACC_Kernels<acc::config<acc::NumVectors<32>>>>> (
    RAJA::RangeSegment{0,32},
    RAJA::RangeSegment{0,32},
    [=] (int out, int in) {
      int idx = out * 32 + in;
      c[idx] = a[idx] + b[idx];
    });

### CODE REVIEW

**Comment by willkill07** [comment]
@trws you were still right about `pgc++` overall. Simple loop nests are fine. SFINAE hell is not.

**Comment by trws** [comment]
If you're emailing them, loop me in on that, I need to bug Doug about some unified memory oddness anyway.  Also, we're supposed to get a significant bump in C++11-14 functionality sometime in July IIRC, it may alleviate some of these pain points.

**Comment by willkill07** [comment]
I'm talking to Mat Colgrove on the OpenACC Slack page. Good news:

> I was able to reproduce the error with 17.4 and 17.5.  The ICE does not occur in our Development compiler but see the following warning and errors.  My best guess is that you have some exception handling in the device code which is causing the problem.  Is there a way to disable exception handling in RAJA?  It gives an error when I use the compiler flag "--no_exceptions".

I'm going to `s/throw .*/assert(false)/` and have him try again.

**Comment by willkill07** [comment]
Here's the full conversation log:

```
% pgc++ -I../../../include -I. -std=c++14 -acc -ta=tesla:managed test.cpp -Vdev
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA11BaseSegmentE (test.cpp: 75)
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA11BaseSegmentE (test.cpp: 84)
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA12RangeSegmentE (test.cpp: 107)
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA12RangeSegmentE (test.cpp: 115)
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA18ForallN_PolicyPairINS_13acc_loop_execINS_3acc6configIINS2_11independentEEEEEENS_12RangeSegmentEEE (test.cpp: 17)
PGCC-W-0155-External and Static variables are not supported in acc routine - _ZTVN4RAJA18ForallN_PolicyPairINS_13acc_loop_execINS_3acc6configIINS2_11independentEEEEEENS_12RangeSegmentEEE (test.cpp: 17)
PGCC-S-1000-Call in OpenACC region to procedure '__cxa_begin_catch' which has no acc routine information (test.cpp: 17)
PGCC-W-0155-Compiler failed to translate accelerator region (see -Minfo messages): Missing branch target block (test.cpp: 1)
PGCC-W-0155-acc loop vector clause ignored in acc routine seq procedure  (test.cpp: 1289)
PGCC-W-0155-acc loop gang clause ignored in acc routine seq procedure  (test.cpp: 1289)
```

> I was able to reproduce the error with 17.4 and 17.5.  The ICE does not occur in our Development compiler but see the following warning and errors.  My best guess is that you have some exception handling in the device code which is causing the problem.  Is there a way to disable exception handling in RAJA?  It gives an error when I use the compiler flag "--no_exceptions".

```
% pgc++ -I../../../include -I. -std=c++14 -acc -ta=tesla:managed test.cpp -V17.4 -w --no_exceptions
"../../../include/RAJA/util/defines.hpp", line 99: error: support for exception
          handling is disabled
          throw std::runtime_error(str);
          ^
1 error detected in the compilation of "test.cpp".
```

> Here's the RAJA error when I try to disable exceptions.

> Thanks.  ICE still occurs with the release, but the warnings went away in the Dev.  Now I'm getting a new one.

```
% pgc++ -I../../../include -I. -std=c++11 -acc -ta=tesla:managed test.cpp -Vdev -w --no_exceptions
PGCC-S-1000-Call in OpenACC region to procedure '_ZNK4RAJA16ForallN_ExecutorIINS_18ForallN_PolicyPairINS_13acc_loop_execINS_3acc6configIINS3_11independentEEEEEENS_12Range...' which has no acc routine information (test.cpp: 107)
PGCC/x86 Linux Rel Dev-r131205: compilation completed with severe errors
```

*Me:* Is there any way to have routine information automatically added? What if I don't know what routine attribute I should be using?

*Me:* really, the goal is for those function calls to be force inlined so routine doesn't need specified

> If the definition is visible to the compiler at compilation, it should auto-create the device routine.  In this case, my guess is that the definition isn't available.   Let me try enabling inlining.

> No luck with inlining.  Must be in a different file.  The mangled name is too long so I'm not sure exactly which routine it's complaining about.  Let me see what I can do to figure it out.

*Me:* I'm pretty sure the function is in `include/RAJA/pattern/forallN.hpp`

The code gets expanded as follows:

`forallN` -> `funUnpacker` -> `forallN_impl` -> `forallN_impl_extract` (which creates a function object in `ForallN_IndexTypeConverter`) -> `forallN_policy` (which creates a function object in `ForallN_Executor`)

`ForallN_Executor` is where all of the loop peeling is done.

The goal is to have `forallN` through `forallN_impl_extract` cleanly inline (since nothing is emitted for a device yet. After this, the outermost directive (`#pragma acc kernels...`) should be emitted with each of the nested execution policies peeled off inside with `forallN_policy`.

Each `forallN_policy` peels off the outer-most policy-range pair and should inline code which then emits a `#pragma acc loop` directive. This is done recursively until all policy-range pairs have been emitted. At this point, it just calls the lambda.

A thought: should I need to annotate the lambda with anything? Do I need to say `[] (int i, int j) __device__ { ... }` ?

**Comment by willkill07** [comment]
TL;DR -- this won't work until June at the earliest regardless. But if I can give enough of a base to have this example working with the compiler release then, that would be really nice. @trws Any further communication with Mat will be over email and I'll CC you. Is there anyone else who I should add?

**Comment by DavidPoliakoff** [comment]
You could try to hit up the OpenACC community more broadly (technical@openacc.org), if you wanted. No pressure to, but I've seen some lively email chains there

**Comment by willkill07** [comment]
Mat's not done looking at it, but I do have a small update:

> I have to head into a series of meetings for a few hours but will get back to this later.  My current thinking is that there are multiple things.

> First, the ICE.  I'll need to report it to our compiler engineers to see why it's happening in the Release but not the Dev.  Was it intentionally fixed or a side-effect of something else and what needs to be done to get it working in a Release

> Second, my current guess is that the missing device routine is the Lambda itself.  The compiler will auto generate device codes for Lambdas but only when it knows it's within an OpenACC compute region.  Here it's being passed into a region.  Need to prove this theory by changing the Lambda to a functor so I can decorate the routine.  We also have a flag "-acc=routineseq" which has the compiler generate device code for every routine in the program, but that has it's own set of problems since not all the routines can be put on the device.

**Comment by willkill07** [comment]
@DavidPoliakoff I think this has more to do with PGI compiler than it does with OpenACC. Theoretically, the issues we are having should be resolved with enough inlining.

**Comment by DavidPoliakoff** [comment]
@willkill07 fair enough, I just know that the general idea of embedding OpenACC would be of interest to them, I've seen conversations with them and Kokkos folks. I'd imagine you could get feedback on why everything we're doing is wrong in technical and moral senses (they're not a shy bunch)

**Comment by cbergstrom** [comment]
Please do not put PIG hacky workarounds for OpenACC implementation stuff. It doesn't help OpenACC, PIG customers or world. I'll try to run the testcase and see what I come up with. If you feel there's a standards issue at hand I don't mind being an objective go-between if needed.
#notshy

**Comment by willkill07** [comment]
@cbergstrom:

0). s/PIG/PGI/g
1). PGI's compiler ICE's while GCC (with -fopenacc) does not, so there is clearly a bug in their compiler. I know a few guys on the PGI compiler team because of my involvement with OLCF GPU Hackathons. I reported the error.
2). This is still in really early preliminary stages. It's something that I'm personally interested in.

> It doesn't help OpenACC

I'm not proposing that pgc++ is the only compiler that I want this to work with. I'm only in discussion with PGI because of an ICE. A conforming modern C++ compiler that also supports OpenACC does help OpenACC.

Regardless, I simplified the test so I get a single error on GCC 6.3 (`error: OpenACC loop directive must be associated with an OpenACC compute region`). I checked how gcc detects compute regions and it's a simple parent loop analysis with GIMPLE which isn't enough because the code wouldn't be inlined right away. Compiling without `-fopenacc` just won't emit the directives, but the machinery is in place.

**Comment by trws** [comment]
That all sounds like good progress and feedback.  Did you ever try a __device__ lambda version? With an nvidia backend on the OpenACC that should generate an appropriate callable signature, even if it isn't portable in the long run it might get to whatever the next issue is if there are more.

**Comment by willkill07** [comment]
Just a brief update: I'll have this updated with some modified policies in the near future. Will definitely want this live by SC'17 for the WACCDP workshop so people can mess around with it. OpenACC reducers are on my TODO list before a big update.

**Comment by willkill07** [comment]
Closing this PR due to `forallN` being deprecated in favor of `nested::forall` -- lots of work behind the scenes required. A new branch will be created once this work is done.

Please do not delete this branch (yet)

**Comment by rhornung67** [comment]
@willkill07 you may want to wait until @ajkunen's new nested stuff is in -- about a month-ish from now.

**Comment by willkill07** [comment]
@rhornung67 that was the plan 😊
